RavenDB-11275 - Support Sorting

redknightlois opened the pull request.

redknightlois committed:
RavenDB-11275 - Implementation of SortingMatch

redknightlois committed:
RavenDB-11275 - Pre-pruning results in order to avoid to sort them.

redknightlois committed:
RavenDB-11275 - Support for caching read objects to avoid rehydration when sorting

ayende commented on src/Corax/IndexEntryBuilder.cs:
```diff
@@ -141,7 +129,7 @@ public void Write(int field, ReadOnlySpan<byte> value, long longValue, double do
 
             // Write known field pointer.
-            _knownFieldsLocations[field] = dataLocation | IsTypedMask[sizeof(int)];
+            _knownFieldsLocations[field] = dataLocation | unchecked((int)0x80000000);
```
Can you define constants here, instead of hard coded values?

redknightlois replied:
Yes, I was fighting the JIT there... this is a candidate for a constant.

ayende commented on src/Corax/IndexSearcher.cs:
```diff
@@ -187,6 +196,71 @@ public MultiTermMatch StartWithQuery(string field, string startWith)
+            //           of showing that it is possible to build any custom group of functions. Why would we want to do this instead
+            //           of just building a TComparer, I dont know. But for now the `CustomMatchComparer` can be built like this from
+            //           static functions. 
+            return Create(new SortingMatch<TInner, CustomMatchComparer>(
```
Agreed, that it not something that we'll likely want and is candidate for removal.

ayende commented on src/Corax/Queries/SortingMatch.Comparers.cs:
```diff
@@ -0,0 +1,305 @@
+            {
+                if (typeof(T) == typeof(long))
+                {
+                    return Math.Sign((long)(object)y - (long)(object)x);
+                }
+                else if (typeof(T) == typeof(double))
+                {
+                    return Math.Sign((double)(object)y - (double)(object)x);
```
Why not call `double.CompareTo`, etc?

redknightlois replied:
Good point, will see if it generated the proper code.

ayende commented on src/Corax/Queries/SortingMatch.Comparers.cs:
```diff
@@ -0,0 +1,305 @@
+
+                    if (readX && readY)
+                    {
+                        return comparer.CompareSequence(resultX, resultY);
```
shouldn't this be the other way around?

redknightlois replied:
Originally it was using the same code for both, that is a leftover of that.

ayende commented on src/Corax/Queries/SortingMatch.Comparers.cs:
```diff
@@ -0,0 +1,305 @@
+
+                    if (readX && readY)
+                    {
+                        return comparer.CompareSequence(resultX, resultY);
```
```suggestion
                        return comparer.CompareSequence(resultY, resultX);
```

ayende replied:
Also, why is that a dedicated comparer instead of just applying `*-1` ?

redknightlois replied:
Because we need the type anyways... Will see if it is doable without losing on the codegen side to just use an instance of an `AscendingMatchComparer` inside.

ayende commented on src/Corax/Queries/SortingMatch.cs:
```diff
@@ -0,0 +1,325 @@
+                _fieldId = comparer.FieldId;
+
+                var hash = ArrayPool<Item>.Shared.Rent(CacheSize);
+                for (int i = 0; i < hash.Length; i++)
```
`Array.Clear` would be faster here?

ayende commented on src/Corax/Queries/SortingMatch.cs:
```diff
@@ -0,0 +1,325 @@
+                    ix.Match = (long)((ulong)x | (readX ? 0 : NoMatch));
+                }
+
+                bool readY;
```
Lots of code duplication here, can we do aggresive inline function instead?

ayende commented on src/Corax/Queries/SortingMatch.cs:
```diff
@@ -0,0 +1,325 @@
+        [SkipLocalsInit]
+        private int Fill<W>(Span<long> matches) where W : struct
+        {
+            int take = _take <= 0 ? matches.Length : _take;
```
There are scenarios where users will send `0` as the take value. For example, when they just care about the number of results.
In this case, why bother sorting?

ayende replied:
What happens when Take is greater than Length?

redknightlois replied:
That should not be the case by design. If we allow `Take` to be longer that `Length` then we need an entirely different approach.

ayende replied:
What happens when `Take` is 20K? Or `int.MaxValue` ?

redknightlois replied:
For unconstrained `Take` we need a version that would take all the potential values and sort afterwards. Different implementation of Sort. This implementation is constrained, and therefore more performant than the unconstrained.

ayende commented on src/Corax/Queries/SortingMatch.cs:
```diff
@@ -0,0 +1,325 @@
+            var sorter = new Sorter<long, HashCacheMatchComparer<TComparer, W>>(hashCacheComparer);
+            sorter.Sort(matches.Slice(0, totalMatches));
+
+            Span<long> a = stackalloc long[matches.Length];
+            Span<long> b = stackalloc long[matches.Length];
```
I think we shouldn't be cleaning this up, if `Length` is high (16K), that can end up being *very* large.
Better to have an "arena" allocation and use that instead, no?

redknightlois replied:
Yes, I plan to use a pattern there that essentially is something like:
```
Span<long> a = matches.Length < 4096 ? stackalloc long[matches.Length] : ArrayPool<long>.Shared.Rent(matches.Length); 
```

ayende replied:
No, why bother with the pool, just allocate a thread local buffer and get from it

redknightlois replied:
Yes, that is an implementation detail. Could be that of basically anything we choose to support that.

ayende commented on src/Corax/Queries/SortingMatch.cs:
```diff
@@ -0,0 +1,325 @@
+        private readonly IQueryMatch _inner;        
+        private readonly TComparer _comparer;
+        private readonly int _take;
+        private unsafe struct SequenceItem
```
Same as `UnmanagedSpan`, no?

ayende commented on src/Voron/Data/Containers/Container.cs:
```diff
@@ -457,17 +462,16 @@ public static Item Get(LowLevelTransaction llt, long id)
             }
 
             var container = new Container(page);
-            ref var metadata = ref container.Offsets[OffsetToIndex(offset)];
-            Debug.Assert(metadata.Size != 0);
-
-            return new Item(page, metadata.Offset, metadata.Size);
+            ItemMetadata* metadata = (ItemMetadata*)(container._page.Pointer + offset);
+            Debug.Assert(metadata->Size != 0);
+            return new Item(page, metadata->Offset, metadata->Size);
         }
 
         private Span<byte> Get(int offset)
         {
-            ref var metadata = ref Offsets[OffsetToIndex(offset)];
-            Debug.Assert(metadata.Size != 0);
-            return Span.Slice(metadata.Offset, metadata.Size);
+            ItemMetadata* metadata = (ItemMetadata*)(_page.Pointer + offset);
+            Debug.Assert(metadata->Size != 0);
+            return Span.Slice(metadata->Offset, metadata->Size);
```
there is a perf difference here?

redknightlois replied:
Yes, huge.

redknightlois committed:
RavenDB-11275 - Optimized code for Set Decoding and Accessing.

redknightlois committed:
RavenDB-11275 - AVX2 codepath for the PForDecode

redknightlois committed:
RavenDB-11275 - Codegen improvements.

- Using pointers to avoid struct inefficient codegen on critical sections.
- Optimizing pointer accesses

redknightlois committed:
RavenDB-11275 - On the fly optimization when we know the size of the set.

redknightlois committed:
RavenDB-11275 - Optimization of BinaryMatch


- And/Or optimizations use of pointer arithmetics for more packed assembler.

ayende committed:
RavenDB-11275 - Adding Lucene benchmark

redknightlois committed:
RavenDB-11275 - Removing the optimization to entirely fill the buffer.

The optimization was aimed to do a better use of the buffer, but as it is it ends up hurting performance because it has a multiplier effect downstream.

redknightlois committed:
RavenDB-11275 - Better codegen for the sorting and deduplication. 

- Remove switch in order for the JIT to inline the methods.
- Reorder the branches to better use the knowledge that first buffer is expected to be bigger on average.

redknightlois committed:
RavenDB-11275 - Updated Benchmarks

redknightlois committed:
RavenDB-11275 - Improvements to the Set iteration and fixes.


- Fixed counting of total MultiMatch values after the first iteration

redknightlois committed:
RavenDB-11275 - Improving sorting to just do N + M reads.

In this way sorting by a property will be faster to execute.

redknightlois committed:
RavenDB-11275 - Diminish the code size of long decoding to allow inlining.

The completely unrolled version of the Variable Size Decoder has to many instructions preventing it to be inlined in some places. Therefore, if the numbers are bigger than ints we will loop instead. We will need to implement a Vectorized version of it for a more compact version.

redknightlois committed:
RavenDB-11275 - Vectorized AND 

The implementation of Vectorized AND is a simplified version of the V1 algorithm from the Lemire paper. 
https://arxiv.org/abs/1401.6399

redknightlois committed:
RavenDB-11275 - Fill based access to set iterator.

redknightlois committed:
RavenDB-11275 - Optimizations for vectorized MergeHelper.And

redknightlois commented:
Updated with the latest vectorized operations. @ayende

redknightlois committed:
RavenDB-11275 -  Vectorized TermMatch.AndWith

redknightlois committed:
RavenDB-11275 - Improvements to the decoding process.

redknightlois committed:
RavenDB-11275 - Improvement to codegen of Vectorized AND

redknightlois committed:
RavenDB-11275 - Remove earlier optimization to avoid repetition

Earlier decoding was far slower, therefore this optimization got us a lot of mileage. But currently the `.Fill()` method is far faster than the `.MoveNext()` therefore the simplicity of OR against the AND allow us to cut the cost of multiterm operations massively.

redknightlois commented:
Include latest improvements that pushes sorting cost to roughly 1/4th of the cost in Lucene.

```
|              Method | BufferSize | TakeSize |     Mean |     Error |    StdDev |
|-------------------- |----------- |--------- |---------:|----------:|----------:|
| OrderByRuntimeQuery |       1024 |       16 | 3.356 ms | 0.0122 ms | 0.0114 ms |
| OrderByRuntimeQuery |       2048 |       16 | 2.188 ms | 0.0067 ms | 0.0060 ms |
| OrderByRuntimeQuery |       4096 |       16 | 1.692 ms | 0.0104 ms | 0.0087 ms |
| OrderByRuntimeQuery |      16384 |       16 | 1.223 ms | 0.0032 ms | 0.0026 ms |
```

And the improvements that push the AndOr benchmark below the 3 digits in microseconds. 
```
|                    Method | BufferSize |      Mean |    Error |   StdDev | Code Size |
|-------------------------- |----------- |----------:|---------:|---------:|----------:|
|              RuntimeQuery |         16 | 437.73 us | 1.995 us | 1.769 us |   2,600 B |
| RuntimeQueryOnlyIteration |         16 | 406.00 us | 1.682 us | 1.574 us |   2,216 B |
|               ParserQuery |         16 | 436.47 us | 1.721 us | 1.526 us |   7,980 B |
|  ParserQueryOnlyIteration |         16 | 409.87 us | 3.075 us | 2.401 us |   7,580 B |
|              RuntimeQuery |         64 | 169.24 us | 1.995 us | 1.666 us |   2,600 B |
| RuntimeQueryOnlyIteration |         64 | 143.23 us | 0.608 us | 0.539 us |   2,216 B |
|               ParserQuery |         64 | 174.60 us | 2.531 us | 3.630 us |   7,980 B |
|  ParserQueryOnlyIteration |         64 | 145.56 us | 0.889 us | 0.694 us |   7,580 B |
|              RuntimeQuery |        256 | 101.01 us | 2.008 us | 2.391 us |   2,600 B |
| RuntimeQueryOnlyIteration |        256 |  78.33 us | 0.682 us | 0.638 us |   2,216 B |
|               ParserQuery |        256 |  97.91 us | 0.274 us | 0.214 us |   7,980 B |
|  ParserQueryOnlyIteration |        256 |  79.26 us | 0.593 us | 0.525 us |   7,580 B |
|              RuntimeQuery |       1024 |  79.01 us | 0.302 us | 0.236 us |   2,600 B |
| RuntimeQueryOnlyIteration |       1024 |  61.38 us | 0.800 us | 0.668 us |   2,216 B |
|               ParserQuery |       1024 |  84.32 us | 0.691 us | 0.613 us |   7,980 B |
|  ParserQueryOnlyIteration |       1024 |  59.26 us | 0.279 us | 0.233 us |   7,580 B |
```

redknightlois committed:
RavenDB-11275 - Improvements to OR supporting non temporal writes.

This will allow us to avoid the processor to go look for the value in memory (in case it is not in cache) and becomes a blind write from the point of view of the processor.
